src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs

```diff
@@ -402,11 +404,10 @@ public async Task<IActionResult> ExternalLoginCallback(string returnUrl = null,
             }
             else
             {
-                var email = info.Principal.FindFirstValue(ClaimTypes.Email) ?? info.Principal.FindFirstValue("email");
-
-                if (!string.IsNullOrWhiteSpace(email))
+                var extLoginMapper = _externalLoginMappers.FirstOrDefault(x => x.CanHandle(info));
+                if (extLoginMapper != null)
                 {
-                    iUser = await _userManager.FindByEmailAsync(email);
+                    iUser = await extLoginMapper.GetUserAsync(info);
```

Comment:
Actually, why isn't `FindByLoginAsync()` above enough?

Reply:
Or better GetAsync or FindAsync. The interface name already means that that context is a user and the parameters already specify by what we are getting the user.

Reply:
I'm not disagreeing with the method name. Rather, why can't we just use `_userManager.FindByLoginAsync()` instead? It dawned on me that this is exactly for such non-email mappings, and it's already used a few lines above.

Reply:
Good point. He can also override the UserStore and change the logic if needed. And with that, we won't need the new interface all together. @PiemP is there a reason why can't we use `_userManager.FindByLoginAsync` and override the UserStore if needed or provide a custom implementation of the `IUserLoginStore<IUser>`

Reply:
> 'm not sure if this "second check" over external login data is correct or it's nature: seems a second chance to connect external login data to user account.

It's what I talking about in the issue ... I don't understand the mean of this "second check" or why we force a user (that could be registered by an external login access and could not have a password) to link a login data to an existing account ... for me is unuseful this second check ... you can specify a `claims` to link login data to an existing account: all in the configuration of the external login method. With this method you can relate a user to a login data without specidy a password and, more important, you can control the value to do this relationship.

The "second check" force to have 1 account 1 email.

I hope to have write something understandable 😄.

Use another time `_userManager.FindByLoginAsync` is unuseful: we have use it in before in the same method. Or we remove this "second check" to link  a user to an existing account or we make it "configurable" by allow users to implement its `IExternalLoginMapper` version.

Reply:
> (that could be registered by an external login access and could not have a "local" password) to link a login data to an existing account ... for me is unuseful this second check

By default, we require unique email address. This is something that can be configured using IdentityOptions. And when we require unique email and the external users matches an existing local user, you MUST provider the local password to ensure that you are linking external account to your local account "not someone else's local account."

However, if we allow duplicate emails, then we should not have to link the external account to a local account "which seems to be your case." Instead, we should just create a new account. But in that case, we somehow have to create a unique username. Maybe in that case, we have to prompt the user to create a username which could also be pre-created based on the settings.

Reply:
> you MUST provider the local password to ensure that you are linking external account to your local account "not someone else's local account."

If you allow user to register at the site with an external login provider... there is the possibility to not set a user local password and I don't have a need to force user to set a password if the user use an external login (that I can configure as my need) to register at the site.

> we should not have to link the external account to a local account "which seems to be your case."

I already have a "first check" that use `_userManager.FindByLoginAsync` to link a local user account to an external login one by a `Claims`.  Why a second one that I can't control? I have thinked about the possibilities to do a PR that introduce a way to disable this check but I prefer this way that allow someone to do what he/she want.

> But in that case, we somehow have to create a unique username. Maybe in that case, we have to prompt the user to create a username which could also be pre-created based on the settings.

During a registration from an external login, actually, we can create a username by a custom rule and we can avoid user from change username and we can make the username unique or use a login data (that is an identifier for the federation) as a username. And this process is completely customizable. 

This PR allow dev to have more possibilities (or maintain the default one)... remain to understand if exist something that could be wrong if someone want to have the relation 1 email N users.

Sorry for my errors or if have made some confusion but it's difficult for me to explain in english. Hope to be not inappropriate.

Reply:
Don't worry about your English, most people here are not native speakers :). I'm a bit confused about the whole thing though. Let me try to clarify:

So your problem 1 is that if someone already has a local user account that they previously created via an external login, and they didn't set a local password (because the site's settings didn't require it), then if they try to add a second external login they'll be asked for a password that they can't provide. Do I understand this correctly?

And your problem 2 is that `UserManager.FindByLoginAsync()` doesn't find the user in your use case, what you aim to address with a new service for selecting the user in this PR. Correct?

Problem 1 I think is a valid concern, but something out of the scope of this PR and its issue, but something we can address separately.

Problem 2 can, however, be addressed by implementing a new `IUserStore<IUser>`, overriding/extending [the built-in one](https://github.com/OrchardCMS/OrchardCore/blob/f60b2247d75cfa99f72603a4b5acaca17efd080e/src/OrchardCore/OrchardCore.Users.Core/Services/UserStore.cs), since this is what [`UserManager` uses behind the scenes](https://github.com/dotnet/aspnetcore/blob/main/src/Identity/Extensions.Core/src/UserManager.cs#L910).

However, even before this, the first question is, why is the [default implementation of `FindByLoginAsync()`](https://github.com/OrchardCMS/OrchardCore/blob/f60b2247d75cfa99f72603a4b5acaca17efd080e/src/OrchardCore/OrchardCore.Users.Core/Services/UserStore.cs#L504) not working for you? Shouldn't we rather fix that in your use case the `UserLoginInfo` is not saved and found properly? Because that should work.

Reply:
> So your problem 1 is that if someone already has a local user account that they previously created via an external login, and they didn't set a local password (because the site's settings didn't require it), then if they try to add a second external login they'll be asked for a password that they can't provide. Do I understand this correctly?

correct.

> And your problem 2 is that UserManager.FindByLoginAsync() doesn't find the user in your use case, what you aim to address with a new service for selecting the user in this PR. Correct?

`UserManager.FindByLoginAsync()` work as expected in the "first check". The problem is that I have set a `ProviderKey` that is different than email ... for this reason, in this lines:

https://github.com/OrchardCMS/OrchardCore/blob/f60b2247d75cfa99f72603a4b5acaca17efd080e/src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs#L382-L388

`iUser` is null and doesn't link User to login data (without ask password, and this is correct for my case). The code goes on and start the "second check". If the "second check" find a User with the email value as the same in the login data:  force to link User and login data. Personally, I don't need this "second check":

https://github.com/OrchardCMS/OrchardCore/blob/f60b2247d75cfa99f72603a4b5acaca17efd080e/src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs#L405-L433

This "second check" force something that is unattended for me and for the configuration of my federation: If I want to link login data to User, the "first check" is enought. Here start my doubt about the "second check" because seems redundant. All work well if the external login identifier is email... but if the external login identifier is another data could be wrong to have this "second check" that force people about link they account to a login data.

> Problem 1 I think is a valid concern, but something out of the scope of this PR and its issue, but something we can address separately.

for me is the same

> Problem 2 can, however, be addressed by implementing a new IUserStore<IUser>, overriding/extending [the built-in one](https://github.com/OrchardCMS/OrchardCore/blob/f60b2247d75cfa99f72603a4b5acaca17efd080e/src/OrchardCore/OrchardCore.Users.Core/Services/UserStore.cs), since this is what [UserManager uses behind the scenes](https://github.com/dotnet/aspnetcore/blob/main/src/Identity/Extensions.Core/src/UserManager.cs#L910).

Actually I have solved my problem by create my `IUserStore<IUser>` implementation taking all the code in `UserStore` rewriting `FindByEmailAsync` to return always null. This avoid the "second check". But is a bad way to do something.

Reply:
I have one external login provider that is authorized to register users in orchard core without asking a password. This external login provider don't use email as `ProviderKey`. My provider key is a different data that are specific of this federation (I use this data to generate username).

> So still, the question is, why doesn't it find one for you?

`_userManager.FindByLoginAsync(info.LoginProvider, info.ProviderKey)` doesn't find a User because the `ProviderKey` of **the new authenticated user** is not registered, because doesn't exist an account for **the new authenticated user** but, unfortunately, **the new authenticated user** have the same email of one registered user and `_userManager.FindByEmailAsync(email)` force **the new authenticated user** to link its data to the local account retrieved by  `_userManager.FindByEmailAsync(email)`.

But in my case, the system should allow to register a new local account for **the new authenticated user** because the `ProviderKey` (of **the new authenticated user**) doesn't match anything ... but the system ask about link data to another local account that, probably, is not of the same person.

I want to skip the email check because for my federation system is unuseful.

EXAMPLE: 
external login provider PLUTO, PLUTO identifier ID

USER A (registered in orchard by provider PLUTO)
email: p@p.it
ID: 1

USER B (not registered in orchard, provider PLUTO)
email: p@p.it
ID: 2
  
if B try to register its account in orchardcore
`_userManager.FindByLoginAsync(info.LoginProvider, info.ProviderKey)` doesn't find a user because the B ID is unregistered
`_userManager.FindByEmailAsync(email)` force B to link its data to A ... but they are 2 different people.

Reply:
in this moment we assume that email is a valid value to link login data to a local user account. But this assumption is not valid for all the cases.

If you have multiple login provider you can use `CanHandle` method to decide when apply a different strategy durign the "second check" to handle the login data for a specific provider and allow or not the linking process.

Reply:
I see, thanks for explaining. So your issue is NOT that the user is not found by `ProviderKey` (it's not found, but that's not an issue, because by that ID the user is actually new), rather that the e-mail is not a unique identifier for your users and thus the login with the second external login is incorrectly matched with the local account of the first one.

Taking a step back: why do you have two user accounts with the same e-mail in your external login system? I'm not disputing the validity of the use case but would like to understand where you're coming from, because I've never seen this in the wild.

I don't think introducing a second service to look up users, as in this PR, is a good solution for this or a good approach in general. As Mike pointed out, `IdentityOptions.Users.RequireUniqueEmail` governs if OC assumes unique e-mail addresses. The problem with the `_userManager.FindByEmailAsync(email)` in `AccountController` then is that it always assumes unique e-mails. Then the solution is simply to make that check depend on `IdentityOptions.Users.RequireUniqueEmail` being `true`.

(The improvements in `LinkExternalLogin.cshtml` and `ExternalLoginInfoExtensions` can also remain, but you'll need to remove `IExternalLoginMapper`.)

Reply:
> why do you have two user accounts with the same e-mail in your external login system?

It's a long strange story... I have called it ¨my federation" but in reality is a public administration federation system that  adopt a strange schema and, about a month ago, I have discovered an article that talk about the possibilities to have different people related to the same email. It's also strange for me...but it's a need.

> I don't think introducing a second service to look up users

It's not a second service to look up users ... `_userManager.FindByEmailAsync(email)` it's already in use ... It's a way to create a custom strategy to link login data to local user account... customizable by what you need using the `CanHandle` method and the service order load in the DI context. 

I implement the changes to use `IdentityOptions.Users.RequireUniqueEmail` and remove `IExternalLoginMapper`.

Reply:
Thank you!

To clarify, `UserManager` (and `IUserStore`) _are_ the services to create a custom strategy. It's that we don't need another one.

Reply:
I see `IExternalLoginMapper` more like `IExternalLoginEventHandler` 😄 ... sorry, I don't have undestrand before your point of view about this fix.

Reply:
Thank you!